fqdn: Fix fqdn subsystem correctness issues causing packet drops

odinuge opened the pull request:
tl;dr this fixes a very bad issue in the fqdn subsystem that cause a permanent broken state where certain IPs get stuck in a state where they cannot be used until they are flushed from the fqdn cache and re-added. That is very problematic when the given fqdn keep returning the same IP(s).

This is still work in progress since its a pretty big change, so want to run test to see if its anything very obviously wrong with it.

Implemented together with @sjdot.

This affects v1.15+, and TODO on me to add the correct `Fixes: ` tag.

Comment message
```
    fqdn: Fix fqdn subsystem correctness issues causing packet drops

    To ensure we never leak/fail to upsert identities, we ensure we only add
    IPs that were already present in the local cache.

    The error we are fixing here is the one where the domain is already in
    the FQDN cache, AND a lookup is in flight. If its added to to local
    endpoint cache before the GC ReplaceFromCacheByNames is executed, it
    could result in it being added to the global cache by the GC,
    hence it never being upserted by the dns lookup code. This leads to a
    situation where the toFQDN won't allow egress to the given IP - and the
    only way to recover for the given IP is to wait until its removed from
    the fqdn cache and a new lookup returns the same IP. This is however not
    very likely, since the same domain can continue to return the same IP
    forever, keeping it constantly in the fqdn cache.

    The same partially applies to endpoint created during the GC iteration,
    resulting in the GC evicting some of the IPs upserted to the global
    cache by that endpoint. That situation could also result in a leak where
    its removed from the fqdn cache, but still be leaked in the ipcache.

    This still needs extensive testing to ensure it doesn't uncover other
    issues that rely on the GC upserting new IPs into the global cache - as
    those could now become even more problematic. We have been running a
    version of this in v1.15 and v1.16 with great success, and without this
    patch we see loads of issues and packet drops due to this issue. We have
    not been testing this in high scale in newer versions than those
    mentioned, and due to other changes in the fqdn subsystem, we need to
    sanity check this thoroughly.

    This is still in works, and will be split into multiple commits, and
    potentially multiple PRs.
```



This also has some improvements in terms of performance as well using the existing `UpdateGenerateDNS` benchmark;
```diff
# based on: go test -bench ^BenchmarkUpdateGenerateDNS$ ./ -benchmem -v -benchtime=20x -count=10
                     │   /tmp/old   │            /tmp/new            │
                     │    sec/op    │    sec/op     vs base          │
UpdateGenerateDNS-12   332.2m ± 20%   356.1m ± 13%  ~ (p=0.529 n=10)

                     │   /tmp/old   │               /tmp/new               │
                     │     B/op     │     B/op      vs base                │
UpdateGenerateDNS-12   5.363Mi ± 2%   4.418Mi ± 2%  -17.62% (p=0.000 n=10)

                     │  /tmp/old   │              /tmp/new              │
                     │  allocs/op  │  allocs/op   vs base               │
UpdateGenerateDNS-12   46.10k ± 0%   42.20k ± 0%  -8.46% (p=0.000 n=10)
```

By patching in another case in the benchmark to show FQDNs with a lot of IPs, the diff is even more evident;
```diff
diff --git a/pkg/fqdn/namemanager/bench_test.go b/pkg/fqdn/namemanager/bench_test.go
index dc5a08cef8..8ffb8e20de 100644
--- a/pkg/fqdn/namemanager/bench_test.go
+++ b/pkg/fqdn/namemanager/bench_test.go
@@ -80,6 +80,11 @@ func BenchmarkUpdateGenerateDNS(b *testing.B) {
                        TTL: 60,
                        IPs: []netip.Addr{ip},
                })
+
+               nameManager.UpdateGenerateDNS(context.Background(), t, dns.FQDN("example.com"), &fqdn.DNSIPRecords{
+                       TTL: 60,
+                       IPs: []netip.Addr{ip},
+               })
        }
 }
```

```
# based on: go test -bench ^BenchmarkUpdateGenerateDNS$ ./ -benchmem -v -benchtime=20x -count=10
                     │ /tmp/old-with-change │         /tmp/new-with-change         │
                     │        sec/op        │    sec/op     vs base                │
UpdateGenerateDNS-12          1116.9m ± 66%   350.4m ± 14%  -68.63% (p=0.000 n=10)

                     │ /tmp/old-with-change │         /tmp/new-with-change         │
                     │         B/op         │     B/op      vs base                │
UpdateGenerateDNS-12         471.672Mi ± 0%   5.516Mi ± 3%  -98.83% (p=0.000 n=10)

                     │ /tmp/old-with-change │        /tmp/new-with-change         │
                     │      allocs/op       │  allocs/op   vs base                │
UpdateGenerateDNS-12            59.15k ± 0%   53.20k ± 0%  -10.06% (p=0.000 n=10)
```


Fixes: #issue-number

```release-note
fqdn: Fix fqdn subsystem correctness issues causing packet drops and inconsistent ipcache
```

odinuge-work committed:
fqdn: Fix fqdn subsystem correctness issues causing packet drops

To ensure we never leak/fail to upsert identities, we ensure we only add
IPs that were already present in the local cache.

The error we are fixing here is the one where the domain is already in
the FQDN cache, AND a lookup is in flight. If its added to local
endpoint cache before the GC ReplaceFromCacheByNames is executed, it
could result in it being added to the global cache by the GC,
hence it never being upserted by the dns lookup code. This leads to a
situation where the toFQDN won't allow egress to the given IP - and the
only way to recover for the given IP is to wait until its removed from
the fqdn cache and a new lookup returns the same IP. This is however not
very likely, since the same domain can continue to return the same IP
forever, keeping it constantly in the fqdn cache.

The same partially applies to endpoint created during the GC iteration,
resulting in the GC evicting some of the IPs upserted to the global
cache by that endpoint. That situation could also result in a leak where
its removed from the fqdn cache, but still be leaked in the ipcache.

This still needs extensive testing to ensure it doesn't uncover other
issues that rely on the GC upserting new IPs into the global cache - as
those could now become even more problematic. We have been running a
version of this in v1.15 and v1.16 with great success, and without this
patch we see loads of issue and packet drops due to this issue. We have
not been testing this in high scale in newer versions than those
mentioned, and due to other changes in the fqdn subsystem, we need to
sanity check this thoroughly.

This is still in works, and will be split into multiple commits, and
potentially multiple PRs.

Signed-off-by: Odin Ugedal <odin@uged.al>
Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge commented on pkg/fqdn/cache.go:
```diff
@@ -503,6 +549,46 @@ func (c *DNSCache) RemoveKnown(mappings map[netip.Addr][]string) {
+		IPs:            ips,
+	}
+
+	c.mu.Lock()
```
I'm not too sure this "longer" critical section is still needed after some of the other changes, so will look into that.

odinuge-work committed:
fqdn: reduce dnsproxy critical section

This makes the DNS cache upserts a bit smarter by tracking if an IP is
new (upserted), or if the lookup just increases the expiration time of a
set of IPs.

This is important since the dnsproxy has to upsert the new IPs to the
ipcache to ensure they are properly plumbed to the datapath.

By doing this there is no need to dump the full DNS cache for the given
name before and after the change, in order to check for new IPs. This
reduces the critical section, and ensure we are not introducing any
race conditions that can lead to either ipcache leaks or missed IPs.

This also reduces the cpu usage of the fqdn subsystem substantially as
well for workloads that do a lot of dns lookups to domains with a lot of
IPs - since sorting slices with 1k or even 10k elements is not free.

This will make per-endpoint Update in the dnsproxy less likely to return
true on "upserted" - resulting in it running `ForceExpireByNameIP` less often. This
should be fine, since the IPs, if not changed, should have been removed
from the zoombie list when it was upserted. It will still trigger the
endpoint header sync as it did before this change.

Signed-off-by: Odin Ugedal <odin@uged.al>
Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge commented on pkg/fqdn/cache.go:
```diff
@@ -193,29 +198,31 @@ func (c *DNSCache) Update(lookupTime time.Time, name string, ips []netip.Addr, t
+func (c *DNSCache) updateWithEntry(entry *cacheEntry) (bool, bool) {
 	entries, exists := c.forward[entry.Name]
 	if !exists {
-		changed = true
```
I figured out this is a bit confusing, so decided to remove. We already know that if we are adding _something_ here, and the fqdn did not exist in the cache - it will always add one or more entries.

odinuge commented:
/fqdn-perf

squeed commented:
Thanks for the fix! I'll take some time to think it over.

I have a far-out idea: I've _never_ liked the per-endpoint DNS caches. I think they cause more harm than good, as this bug is evidence of. I think they only exist because endpoints have an existing mechanism for writing to disk.

Could you imagine a world where we only have one, global FQDN cache? Persistence could be achieved by using SQLite or even boltdb. Then we could rip this complexity all out and be done with it for good. Plus, we wouldn't have the overhead of writing lots of json just for the endpoint.

odinuge commented on pkg/fqdn/cache.go:
```diff
@@ -178,7 +183,7 @@ func (c *DNSCache) DisableCleanupTrack() {
 // ips may be an IPv4 or IPv6 IP. Duplicates will be removed.
 // ttl is the DNS TTL for ips and is a seconds value.
-func (c *DNSCache) Update(lookupTime time.Time, name string, ips []netip.Addr, ttl int) bool {
+func (c *DNSCache) Update(lookupTime time.Time, name string, ips []netip.Addr, ttl int, updates ...*DNSCache) (bool, bool) {
```
I think I'll update this to return a struct instead to avoid the `(bool, bool)` and potentially messing up the ordering.

odinuge-work committed:
fqdn: Add a GC test for name manager

The name manager was lacking a proper test for its GC behavior, so this
is adding that.

Signed-off-by: Odin Ugedal <odin@uged.al>
Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge-work committed:
fqdn: change where we calculate stale IPs from

This ensures we use the data from immediately before the cleanup when
checking for what IP<>name mappings we can clean up in the ipcache. This
ensure we don't loose some edge-cases where we end up cleaning up IPs
that previously were not present at the time of the maybeStaleIPs
caluclation

Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge-work committed:
fqdn: change function returns to be easier to read

This is a noop that switches from returning two bools with almost the
same name, to instead returning a struct.

Signed-off-by: Odin Ugedal <odin@uged.al>
Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge-work committed:
fqdn: add more checks to FQDN tests

Add some more checks to ensure that the fqdn cache is consistent with
the ipcache.

Signed-off-by: Odin Ugedal <odin@uged.al>
Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge commented:
While working on this I discovered a set of other fqdn related issues in the ipcache. Its effectively impossible to write proper tests and test this e2e without fixing that issue, so lets work on https://github.com/cilium/cilium/pull/42815 first. 😄

odinuge-work committed:
fqdn: rewrite name manager test to use real ipcache

Signed-off-by: Odin Ugedal <odin@uged.al>
Signed-off-by: Odin Ugedal <ougedal@palantir.com>
